snuba/migrations/snuba_migrations/stacktraces/0001_stacktraces.py

```diff
@@ -0,0 +1,95 @@
+from typing import List, Sequence
+
+from snuba.clickhouse.columns import (
+    UUID,
+    Column,
+    DateTime,
+    String,
+    UInt,
+    NamedTuple,
+)
+from snuba.clusters.storage_sets import StorageSetKey
+from snuba.migrations import migration, operations, table_engines
+from snuba.migrations.columns import MigrationModifiers as Modifiers
+
+columns: List[Column[Modifiers]] = [
+    # primary key
+    Column("organization_id", UInt(64)),
+    Column("project_id", UInt(64)),
+    Column("transaction_id", UUID()),
+    Column("received", DateTime()),
+
+    # profiling data
+    Column("stacktrace", String(Modifiers(codecs=["LZ4HC(9)"]))),
+    Column("symbols", String(Modifiers(codecs=["LZ4HC(9)"]))),
+
+    # filtering data
+    Column("trace_id", UUID()),
+    Column("transaction_name", String(Modifiers(low_cardinality=True))),
+    Column("environment", String(Modifiers(nullable=True, low_cardinality=True))),
+
+    Column("version", NamedTuple((("name", String()), ("code", String())))),
+    Column("platform", String(Modifiers(low_cardinality=True))),
+    Column("android_api_level", UInt(32, Modifiers(nullable=True))),
+    Column("device_classification", String(Modifiers(low_cardinality=True))),
+    Column("device_locale", String(Modifiers(low_cardinality=True))),
+    Column("device_manufacturer", String(Modifiers(low_cardinality=True))),
+    Column("device_model", String(Modifiers(low_cardinality=True))),
+    Column("device_os_build_number", String(Modifiers(low_cardinality=True))),
+    Column("device_os_name", String(Modifiers(low_cardinality=True))),
+    Column("device_os_version", String(Modifiers(low_cardinality=True))),
+
+    Column("error_code", String(Modifiers(low_cardinality=True, nullable=True))),
+    Column("error_description", String(Modifiers(low_cardinality=True, nullable=True))),
+
+    Column("duration_ns", UInt(64)),
+
+    # internal data
+    Column("retention_days", UInt(16)),
+]
+
+
+class Migration(migration.ClickhouseNodeMigration):
+    blocking = False
+
+    def forwards_local(self) -> Sequence[operations.SqlOperation]:
+        return [
+            operations.CreateTable(
+                storage_set=StorageSetKey.STACKTRACES,
+                table_name="stacktraces_local",
+                columns=columns,
+                engine=table_engines.MergeTree(
+                    storage_set=StorageSetKey.STACKTRACES,
+                    order_by="(organiation_id, project_id, transaction_id, received)",
```

Comment:
```suggestion
                    order_by="(organization_id, project_id, transaction_id, received)",
```

Reply:
I'm still wondering if I should keep that field. You seem to only work with project IDs anyway so this might be redundant. Thoughts?

Reply:
I don't really have any strong opinion about this. @evanh might have thoughts on how this impacts things on the query side though.

Reply:
I think you should keep `organization_id`. We do work with organization IDs in different parts of the product. In fact the only datasets that don't use organization IDs are errors and transactions.

I don't think there is any downside to having organization ID here, besides using slightly more storage. There should never be a case in the product where you have a project ID with no organization ID, and there are times when you want to query across an entire organization (for example if the user has every project selected).

Reply:
A couple of remarks around the the order by key. 
This serves two purposes: primary key (for deduplication and merge when you will turn this into a ReplacingMergeTree), and sorting key for queries. 
So first you need to ensure that you are adding enough fields that allow you to deduplicate rows otherwise (on a replacing merge tree) clickhouse will replace those that have the same order by key.

Second it is a matter of search efficiency.
- organization and project first are good. You always have them in your query. Generally we do not need to have org id in snuba unless you need it for some reason (querying across an organization without specifying a project. DO you have such use case?)
- Then you probably want the time range. If queries based on time ranges are a thing. From your scenario they are. So you can put `toStartOfDay(timestamp or whatever time column)`. This narrows down the dataset dramatically when doing aggregation queries.
- Then you want the trace_id/transaction_id. Which one do you plan to search by ? 

Whether we should have timestamp first or transaction_id first depends on what do you want to optimize between search by trace or search by time range.
- If you put transaction_id first, Clickhouse will be extremely efficient at finding one trace. Querying by any other field within a time range will be a nightmare. This is because you will have to scan literally the entire table as timestamps would not be sorted. Clickhouse will not be able to exclude any block based on timestamp since the table is sorted by transaciton.
- If you put timestamp first, you will have much more efficient queries by time range. Clickhosue will be able to narrow down the portion of the table to scan very quickly. (so short time range queries will be more efficient).  The search by transaciton_id will be somewhat less effective as you will have to find the trace _id on every day potentially (you are sorting by toStartOfDay(timestamp). That is not a major issue as the sections of the table to search are relatively few (90 days). This can be made better if we knew somehow a shorter time range in which to look for a a transaction.

Reply:
If the queries that extract a single transactions are very prominent, I would suggest to store the traces separately (like in nodestore) as we do for events/transacitons to have an effective search by transaction id.

Reply:
Have you also considered adding the platform (ios/android) in the order by ? Is there any chance you will know whether you are looking for an android trace or an ios trace before querying ? That would exclude all the rows from the other platform.